java-design-patterns-master/reactor/src/main/java/com/iluwatar/reactor/app/AppClient.java

```diff
@@ -0,0 +1,194 @@
+ * THE SOFTWARE.
+ */
+
+package com.iluwatar.reactor.app;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Represents the clients of Reactor pattern. Multiple clients are run concurrently and send logging
+ * requests to Reactor.
+ */
+public class AppClient {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AppClient.class);
+
+  private final ExecutorService service = Executors.newFixedThreadPool(4);
+
+  /**
+   * App client entry.
+   *
+   * @throws IOException if any I/O error occurs.
+   */
+  public static void main(String[] args) throws IOException {
+    var appClient = new AppClient();
+    appClient.start();
+  }
+
+  /**
+   * Starts the logging clients.
+   *
+   * @throws IOException if any I/O error occurs.
+   */
+  public void start() throws IOException {
+    LOGGER.info("Starting logging clients");
+    service.execute(new TcpLoggingClient("Client 1", 6666));
+    service.execute(new TcpLoggingClient("Client 2", 6667));
+    service.execute(new UdpLoggingClient("Client 3", 6668));
+    service.execute(new UdpLoggingClient("Client 4", 6668));
+  }
+
+  /**
+   * Stops logging clients. This is a blocking call.
+   */
+  public void stop() {
+    service.shutdown();
+    if (!service.isTerminated()) {
+      service.shutdownNow();
+      try {
+        service.awaitTermination(1000, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOGGER.error("exception awaiting termination", e);
+      }
+    }
+    LOGGER.info("Logging clients stopped");
+  }
+
+  private static void artificialDelayOf(long millis) {
+    try {
+      Thread.sleep(millis);
+    } catch (InterruptedException e) {
+      LOGGER.error("sleep interrupted", e);
+    }
+  }
+
+  /**
+   * A logging client that sends requests to Reactor on TCP socket.
+   */
+  static class TcpLoggingClient implements Runnable {
+
+    private final int serverPort;
+    private final String clientName;
+
+    /**
+     * Creates a new TCP logging client.
+     *
+     * @param clientName the name of the client to be sent in logging requests.
+     * @param serverPort the port on which client will send logging requests.
+     */
+    public TcpLoggingClient(String clientName, int serverPort) {
+      this.clientName = clientName;
+      this.serverPort = serverPort;
+    }
+
+    @Override
+    public void run() {
+      try (Socket socket = new Socket(InetAddress.getLocalHost(), serverPort)) {
+        var outputStream = socket.getOutputStream();
+        var writer = new PrintWriter(outputStream);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *writer*. Currently, there are execution paths that do not contain closure statements. Either a) close *writer* in a try-finally block or b) close the resource by declaring *writer* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).